core/controllers/reader.py

```diff
@@ -1274,3 +1299,103 @@ def put(self, entity_type, entity_id):
             entity_type, state_reference,
             interaction_id, answer, answer_details)
         self.render_json({})
+
+
+class CheckpointCompletedEventHandler(base.BaseHandler):
+    """Tracks a learner completing a checkpoint."""
+
+    URL_PATH_ARGS_SCHEMAS = {
+        'exploration_id': {
+            'schema': editor.SCHEMA_FOR_EXPLORATION_ID
+        }
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'POST': {
+            'version': {
+                'schema': editor.SCHEMA_FOR_VERSION
+            },
+            'state_name': {
+                'schema': {
+                    'type': 'basestring',
+                    'validators': [{
+                        'id': 'has_length_at_most',
+                        'max_value': constants.MAX_STATE_NAME_LENGTH
+                    }]
+                }
+            },
+        }
+    }
+
+    REQUIRE_PAYLOAD_CSRF_CHECK = False
+
+    @acl_decorators.can_play_exploration
+    def post(self, exploration_id):
+        """Handles POST requests.
+
+        Args:
+            exploration_id: str. The ID of the exploration.
+        """
+
+        user_id = self.user_id
+        last_completed_checkpoint_state_name = self.normalized_payload.get(
+            'state_name')
+        last_completed_checkpoint_exp_version = self.normalized_payload.get(
+            'version')
+
+        user_services.set_last_completed_checkpoint(
```

Comment:
The name of this function plus the variables passed seams misleading as we are not just setting last completed checkpoint but also setting latest visited checkpoint in this case. It would be better to change it's name to something like update_learners_checkpoint_progress() or something similar.

Reply:
Done!